### PR TITLE
Kleisli Composition Operator and Apply_or Added

### PR DESCRIPTION
**Pull request by NoahBatchelor**
Added the Kleisli composition operator for Option, Result, and CCFun.

### CODE REVIEW

File: src/core/CCResult.mli [hunk 1]
```diff
@@ -199,6 +203,11 @@ module Infix : sig
 
   val ( and* ) : ('a, 'e) t -> ('b, 'e) t -> ('a * 'b, 'e) t
   (** @since 2.8 *)
+
+  val ( >=> ) : ('a -> ('b, 'err) t) -> ('b -> ('c, 'err) t) -> ('a -> ('c, 'err) t)
+  (** Monadic [k_compose]. *)
```
Comment (c-cube) [inline src/core/CCResult.mli:208]:
```suggestion
  (** Monadic [k_compose]. *)
  
```

File: src/core/CCOption.mli [hunk 2]
```diff
@@ -179,6 +182,11 @@ module Infix : sig
   val ( and+ ) : 'a t -> 'b t -> ('a * 'b) t
   val ( let* ) : 'a t -> ('a -> 'b t) -> 'b t
   val ( and* ) : 'a t -> 'b t -> ('a * 'b) t
+
+  val ( >=> ) : ('a -> 'b t) -> ('b -> 'c t) -> ('a -> 'c t)
+  (** Monadic [k_compose]. *)
```
Comment (c-cube) [inline src/core/CCOption.mli:187]:
```suggestion
  (** Monadic [k_compose]. *)
  
```

File: src/core/CCFun.mli [hunk 3]
```diff
@@ -105,8 +105,18 @@ end) : sig
   val return : 'a -> 'a t
   (** Monadic [return]. *)
 
+  val k_compose : ('a -> 'b t) -> ('b -> 'c t) -> ('a -> 'c t)
+  (** Kleisli composition. Monadic equivalent of [compose]. *)
+
   val ( >|= ) : 'a t -> ('a -> 'b) -> 'b t
+  (** Mondaic [map]. *)
```
Comment (c-cube) [inline src/core/CCFun.mli:112]:
```suggestion
  (** Monadic [map]. *)
```

**Comment by NoahBatchelor** [comment]
Updated PR to remove CCFun changes, added slight documentation to CCFun, and added a new apply_or function.

File: src/core/CCOption.ml [hunk 4]
```diff
@@ -111,6 +116,13 @@ let get_or ~default x =
   | None -> default
   | Some y -> y
 
+let apply_or f x =
```
Comment (c-cube) [inline src/core/CCOption.ml:119]: isn't that `fold`?

Comment (NoahBatchelor) [reply src/core/CCOption.ml:118]: It was motivated mostly by seeing places where CCString functions were being stringed together, combined with repeated CCOption's `get_or` to reset back to the original string if the string functions failed. Particularly the operator form would concisely shrink down that sort of pipeline to just the string functions.

**Pull request merged**